### PR TITLE
Add support for SPV_INTEL_fpga_buffer_location extension

### PR DESCRIPTION
**Pull request by MrSidims**
This extension adds a function parameter decoration that is useful
for FPGA targets. This decoration indicates that a particular global
memory pointer can only access a particular physical memory location.
Knowing this information at compile time can allow FPGA compilers to
generate load store units of lower area for accesses done through such
a pointer.

Specification: https://github.com/intel/llvm/pull/2084

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

### CODE REVIEW

**Comment by mlychkov** [comment]
In a commit message: "memory pointer can only access a _**articular**_ physical memory location."

File: test/transcoding/FPGABufferLocation.ll [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+; CHECK-LLVM: define spir_kernel void @test{{.*}} !kernel_arg_buffer_location ![[BUFLOC_MD:[0-9]+]] {{.*}}
+; CHECK-LLVM: ![[BUFLOC_MD]] = !{i32 1, i32 2, i32 0}
+
+; ModuleID = 'buffer_location.cl'
+source_filename = "buffer_location.cl"
+target datalayout = "e-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024"
+target triple = "spir64-unknown-unknown"
+
+; Function Attrs: norecurse nounwind readnone
+define spir_kernel void @test(i32 addrspace(1)* %a, float addrspace(1)* %b, i32 addrspace(1)* %c) local_unnamed_addr !kernel_arg_addr_space !3 !kernel_arg_access_qual !4 !kernel_arg_type !5 !kernel_arg_base_type !5 !kernel_arg_buffer_location !6 {
```
Comment (mlychkov) [inline test/transcoding/FPGABufferLocation.ll:35]: It would be nice to have also non-pointer parameter to see how it is (not) processed.

Comment (MrSidims) [reply test/transcoding/FPGABufferLocation.ll:35]: I will add this case.
IMO non-pointer parameters shall also be processed (as it's done in this implementation), I'll raise a spec issue.

Comment (MrSidims) [reply test/transcoding/FPGABufferLocation.ll:35]: I got the answer, guess, I need to adjust the implementation.

File: lib/SPIRV/libSPIRV/spirv.hpp [hunk 2]
```diff
@@ -956,6 +957,7 @@ enum Capability {
   CapabilityKernelAttributesINTEL = 5892,
   CapabilityFPGAKernelAttributesINTEL = 5897,
   CapabilityUSMStorageClassesINTEL = 5935,
+  CapabilityFPGABufferLocationINTEL = 5920,
```
Comment (MrSidims) [inline lib/SPIRV/libSPIRV/spirv.hpp:960]: To align

File: test/transcoding/FPGABufferLocation.ll (hunk 1 above)
Comment (AlexeySotkin) [reply test/transcoding/FPGABufferLocation.ll:35]: If this PR is not ready for merge convert it to draft.

Comment (MrSidims) [reply test/transcoding/FPGABufferLocation.ll:35]: Done

Comment (MrSidims) [reply test/transcoding/FPGABufferLocation.ll]: Expanded the test

**Comment by MrSidims** [comment]
Rebased and amended the commits. Also fixed a typo in the commit message.

**Pull request merged**